client/hooks/use-report-export.ts

```diff
@@ -0,0 +1,177 @@
+ */
+import { useEffect, useRef, useState } from 'react';
+import apiFetch from '@wordpress/api-fetch';
+import { useDispatch } from '@wordpress/data';
+import { __, sprintf } from '@wordpress/i18n';
+
+interface ExportResponse {
+	export_id?: string;
+}
+
+const maxRetries = 5;
+
+/**
+ * Hook for requesting and polling for a CSV export. E.g. Transactions, Payouts, Disputes.
+ *
+ * @example
+ * const { requestReportExport, isDownloading } = useReportExport();
+ * requestReportExport( {
+ * 	exportRequestURL: '/wc/v3/payments/transactions/download?queryParam=value',
+ * 	exportFileAvailabilityEndpoint: '/wc/v3/payments/transactions/download',
+ * 	userEmail: 'test@example.com',
+ * } );
+ */
+export const useReportExport = () => {
+	const [ isDownloading, setIsDownloading ] = useState( false );
+	const { createNotice } = useDispatch( 'core/notices' );
+	const timeoutIdRef = useRef< NodeJS.Timeout | null >( null );
+	const retryCountRef = useRef( 0 );
+
+	useEffect( () => {
+		// Cancel the timeout if the component unmounts.
+		return () => {
+			if ( timeoutIdRef.current ) {
+				clearTimeout( timeoutIdRef.current );
+			}
+		};
+	}, [ timeoutIdRef ] );
+
+	interface PollForFileProps {
+		/**
+		 * The URL to check for the exported file. E.g. "/wc/v3/payments/transactions/download/{export_id}"
+		 */
+		checkFileURL: string;
+		/**
+		 * The email address the export will be sent to.
+		 */
+		userEmail: string;
+		/**
+		 * The interval in milliseconds to poll the server.
+		 */
+		interval?: number;
+	}
+
+	/**
+	 * Polls the server to check if the exported file is ready.
+	 */
+	async function pollForFile( {
+		checkFileURL,
+		userEmail,
+		interval = 1000,
+	}: PollForFileProps ) {
+		timeoutIdRef.current = setTimeout( async () => {
+			retryCountRef.current++;
+			const exportedFileURL = await apiFetch< string >( {
+				path: checkFileURL,
+				method: 'GET',
+			} );
+
+			if ( exportedFileURL ) {
+				// The file is available, so we can download it.
+				// Create a link element to trigger the download.
+				const link = document.createElement( 'a' );
+				// Add force_download=true to the URL to force the download, which adds the appropriate `Content-Disposition: attachment` header when using production server.
+				link.href = exportedFileURL + '?force_download=true';
+				link.click();
+
+				createNotice(
+					'success',
+					sprintf(
+						__(
+							'Your CSV export is ready to download. It will also be emailed to %s',
+							'woocommerce-payments'
+						),
+						userEmail
+					)
+				);
+
+				setIsDownloading( false );
+
+				return;
+			}
+
+			if ( retryCountRef.current < maxRetries ) {
+				// If the file is not available, check again after 1 second.
+				pollForFile( {
+					checkFileURL,
+					userEmail,
+				} );
+			} else {
+				// If the file is not available after the maximum number of retries, show an error notice.
+				setIsDownloading( false );
+				createNotice(
+					'error',
+					__(
```

Comment:
@Jinksi  It just means the job is slow / URL not available. At this point, it should say "Your export will be emailed .."